docs/src/rfcs/002-fearless-concurrency.md

```diff
@@ -0,0 +1,191 @@
+
+#### Borrowing
+
+Borrowing is a staple of rust ownership - but in order to be explicit about the scope at which values are visible, the user has to declare this intent when borrowing. Rather than a simple "borrow" and "borrow_mut" method, `Buffer` instead has the following interfaces:
+
+```rust
+fn borrow_for_device(&self) -> DeviceVisibleBufferValue<T>;
+fn borrow_mut_for_device(&mut self) -> DeviceVisibleMutableBufferValue<T>;
+fn borrow_for_workgroup(&self) -> WorkgroupVisibleBufferValue<T>;
+fn borrow_mut_for_workgroup(&mut self) -> WorkgroupVisibleMutableBufferValue<T>;
+fn borrow_for_subgroup(&self) -> SubgroupVisibleBufferValue<T>;
+fn borrow_mut_for_subgroup(&mut self) -> SubgroupVisibleMutableBufferValue<T>;
+```
+
+Note that the borrow functions don't directly return a reference to the underlying value - instead they return intermediate types which mediate correct loads and stores, described in the following section.
+
+For each borrow, an `OpMemoryBarrier` instruction is emitted with a `Scope` equal to that of the name of the borrow command, and `Acquire | UniformMemory` memory semantics.
+
+#### Scope Visible Wrappers
+
+Rust doesn't cater for different "types" of memory or annotation at the moment, and adding that into the compiler would be a significant amount of work. It might be possible in future to use some sort of annotated type, but for now this is going down a similar route to that of the `Atomic*` types in the standard library by wrapping underlying variables with types that have dedicated load/store methods.
+
+Each wrapper type is named with its scope (Device/Workgroup/Subgroup), and whether it is mutable or not, reflecting its underlying properties.
+
+##### Loads and Stores
+
+Loads and stores are dedicated methods on the commands returned from the `Buffer` borrows. All of those types have a `load()` method; only the `Mutable` types include the `store()` method.
+
+```rust
+fn load(&self) -> T;
+fn store(&mut self, value: T);
+```
+
+The two methods mostly map exactly to what you'd expect - `OpLoad` and `OpStore`, however each of them will include additional `Memory Operands`.
+
+For `load()`, an `OpLoad` instruction is generated with the `MakePointerVisible` semantic and a following `Scope` parameter equal to the scope in the name of the wrapper type.
+
+For `store()`, an `OpStore` instruction is generated with the `MakePointerAvailable` semantic, and a following `Scope` parameter equal to the scope in the name of the wrapper type.
+
+##### Drop
+
+The `Mutable` variants of the wrapper types implement a custom `drop()` to ensure correct synchronization.
+
+```rust
+fn drop(&mut self);
+```
+
+In addition to destruction when it goes out of scope, an `OpMemoryBarrier` instruction is emitted with a `Scope` equal to that of the scope of the wrapper type, and `Release | UniformMemory` memory semantics.
```

Comment:
As others mentioned on Discord, [you can't rely on `drop` being run for safety.](https://doc.rust-lang.org/reference/destructors.html#not-running-destructors) If you want to guarantee destruction and replicate Rust's ownership semantics I would use closures. With closures you can pass it temporarily or give it to the closure but require that the user returns it.

```rust
struct Buffer;

impl Buffer {
    // Version A (mutable ref)
    fn borrow_mut_for_device(&mut self, callback: impl FnOnce(&mut DeviceVisibleMutableBufferValue)) {
        let buffer_value = _; // create DeviceVisibleMutableBufferValue
        callback(&mut buffer_value);
        self.destroy(buffer_value);
    }

    // Version B (passing ownership)
    fn borrow_mut_for_device(&mut self, callback: impl FnOnce(DeviceVisibleMutableBufferValue) -> DeviceVisibleMutableBufferValue) {
        let buffer_value = _; // create DeviceVisibleMutableBufferValue
        let buffer_value = callback(buffer_value);
        self.destroy(buffer_value);
    }
}

fn main() {
    // Version A
    buffer.borrow_mut_for_device(|buffer_value| {
        let value = buffer_value.load();
        // Use value...
    })
    // Version B
    buffer.borrow_mut_for_device(|buffer_value| {
        let value = buffer_value.load();
        // Use value...
        buffer_value
    })
}
```

Reply:
Ok yea with fresh eyes I now see why it's considered "safe" to effectively leak objects - link was helpful too, thanks!

Closures probably are a decent option for at least more advanced use of buffers - though "simple" load/store might suffer here. Going to have a think about this one, but it's given me some good ideas at least.

Reply:
Right, I've taken this feedback on board and flipping run with it - I've updated the proposal to V2.